torch/testing/_internal/distributed/multi_threaded_pg.py

```diff
@@ -140,6 +143,8 @@ def __init__(self, world_size, collective):
         self._count = 0
         self._done = False
 
+        self._pg = pg
```

Comment:
I believe the terminate flag should be stored at PG level because it is associated with each test rather than each collective operation. Mostly the error in a thread happens between collective operations like:
`first collective ends ------- AssertionError -------> second collective starts`
and `_start_coll` function is called at the start of every collective to assign a new Collective object to PG:
`cls._cur_coll = Collective(world_size, collective, cls)`

Therefore, if we store the flag in Collective, this information will be stored in the first Collective object when AssertionError happens, and will be lost when the second Collective object is assigned to `_cur_coll` when the second collective starts.

It's true that we can declare the flag as a static field within Collective but it seems against the original design of Collective class.

What do you think?

Reply:
Why ProcessGroup level associate with each test? each test could have multiple process groups right? I don't think ProcessGroup and test have 1:1 correspondance. Even though some of the threads might not have the `self._cur_coll`,  does it really matter as what we really want to do is to turn off the collective that is hanging right? We only need to signify the threads that have `self._cur_coll` and throw an exception in that thread?

Reply:
When something like AssertionError occurs, that thread may have `self._cur_coll = None` because at that moment the collective already completes and `_end_coll()` sets `_cur_coll` to `None`.